go/project_structure/library.md

```diff
@@ -0,0 +1,113 @@
+# Library
+
+This page describes common code structure for our libraries.
+
+The example below is taken from [k8sclient] on 2020/04/16. Most of the files
+and directories are elided for simplification.
+
+```
+github.com/giantswarm/k8sclient/
+├── CHANGELOG.md
+├── README.md
+├── go.mod
+├── go.sum
+└── pkg
+    ├── k8sclient
+    │   ├── clients.go
+    │   ├── error.go
+    │   ├── setup.go
+    │   ├── spec.go
+    │   ├── spec_test.go
+    │   └── types.go
+    ├── k8sclienttest
+    │   └── clients.go
+    ├── k8scrdclient
+    │   ├── crd_client.go
+    │   ├── crd_client_test.go
+    │   ├── error.go
+    │   └── spec.go
+    ├── k8srestconfig
+    │   ├── error.go
+    │   └── rest_config.go
+    └── k8sversion
+        ├── error.go
+        ├── funcs.go
+        ├── funcs_test.go
+        └── types.go
+```
+
+Below there are a few packages discussed.
+
+## Root package
+
+Note that since we use Go modules we don't have any Go files in the root of the
+repository. This is because as we bump the major version of the module we need
+to use `/vN/` imports. We put everything under subpackages of `/pkg` to avoid
+some ugly import aliases like:
+
+```
+k8sclientv3 "github.com/giantswarm/k8sclient/v3"
+```
+
+Instead the import will look like:
+
+```
+"github.com/giantswarm/k8sclient/v3/pkg/k8sclient"
+```
+
+This comes at expense of longer imports of other packages. E.g. instead of
+having import like:
+
+```
+"github.com/giantswarm/k8sclient/v3/k8scrdclient"
+```
+
+We have import like:
+
+```
+"github.com/giantswarm/k8sclient/v3/pkg/k8scrdclient"
+```
+
+Which seems to be a fair price to pay.
+
+## k8scrdclient
+
+```
+./pkg/k8scrdclient
+├── crd_client.go
+├── crd_client_test.go
+├── error.go
+└── spec.go
+```
+
+- `crd_client.go` holds `CRDClient` type implementation and `Config` type it is
+  configured with.
+- `crd_client_test.go` holds tests for `CRDClient` type.
+- `error.go` holds error definitions for the package.
+- `spec.go` holds `Interface` type which is interface for this package (in this
+  case interface which `CRDClient` type implements).
+
+This is a typical layout for a package holding a single type providing methods
+executing some business logic. In this case `*k8scrdclient.CRDClient`. We often
+have corresponding test package, e.g. `k8scrdclienttest` implementing a mock
+object implementing `Interface` type.
+
+## k8sversion
+
+```
+./pkg/k8sversion
+├── error.go
+├── funcs.go
+├── funcs_test.go
+└── types.go
+```
+
+- `error.go` holds error definitions for the package.
+- `funcs.go` holds functions which are not methods of any struct (in this case
```

Comment:
I have to say I'm not personally super thrilled to organize code into files this way. I prefer to have file per context and combine related stuff there, but I might be minority with my opinion.

Reply:
I didn't mean enforcing here. I rather wanted to provide a default. I understand this doesn't fit everything. That's why I wrote `This is a typical layout`.

TBH it started with `/pkg` and Go modules `vN`. I'm even fine with removing whole `k8scrdclient` section if it feels "offensive".

Reply:
+1 on not enforcing it. The default guideline makes sense to me.

Reply:
Sorry I was a bit ambiguous here. I meant specifically `funcs.go` (like we similarly have `error.go`, `spec.go` and `types.go`).

I perfectly understand the `/pkg` structure with Go modules `vN` approach. I'm just converting `azure-operator` to Go modules and already hit this specific import "issue" few times.

Having a guideline is good. Strict enforcing not so :slightly_smiling_face: